archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,572 @@
+  refine hBQ₁Q _,
+  rw [←h, set.pair_comm, set.insert_comm],
+  exact cfg.sbtw_Q_A₁_Q₁.wbtw.collinear
+end
+
+lemma sbtw_B_A₁_C : sbtw ℝ cfg.B cfg.A₁ cfg.C := ⟨cfg.wbtw_B_A₁_C, cfg.A₁_ne_B, cfg.A₁_ne_C⟩
+
+lemma sbtw_A_B₁_C : sbtw ℝ cfg.A cfg.B₁ cfg.C := cfg.symm.sbtw_B_A₁_C
+
+lemma sbtw_A_A₁_A₂ : sbtw ℝ cfg.A cfg.A₁ cfg.A₂ :=
+begin
+  refine sphere.sbtw_second_inter cfg.A_mem_circumsphere _,
+  convert cfg.sbtw_B_A₁_C.dist_lt_max_dist _,
+  change _ = max (dist (cfg.triangle_ABC.points 1) _) (dist (cfg.triangle_ABC.points 2) _),
+  simp_rw [circumsphere_center, circumsphere_radius, dist_circumcenter_eq_circumradius, max_self]
+end
+
+lemma sbtw_B_B₁_B₂ : sbtw ℝ cfg.B cfg.B₁ cfg.B₂ :=
+by { rw ←cfg.symm_A₂, exact cfg.symm.sbtw_A_A₁_A₂ }
+
+lemma A₂_ne_A : cfg.A₂ ≠ cfg.A := cfg.sbtw_A_A₁_A₂.left_ne_right.symm
+
+lemma A₂_ne_P : cfg.A₂ ≠ cfg.P := (cfg.sbtw_A_A₁_A₂.trans_wbtw_left_ne cfg.wbtw_A_P_A₁).symm
+
+lemma A₂_ne_B : cfg.A₂ ≠ cfg.B :=
+begin
+  intro h,
+  have h₁ := cfg.sbtw_A_A₁_A₂,
+  rw h at h₁,
+  refine cfg.not_collinear_ABC _,
+  have hc : collinear ℝ ({cfg.A, cfg.C, cfg.B, cfg.A₁} : set Pt) :=
+    collinear_insert_insert_of_mem_affine_span_pair h₁.left_mem_affine_span
+      cfg.sbtw_B_A₁_C.right_mem_affine_span,
+  refine hc.subset _,
+  rw [set.pair_comm _ cfg.A₁, set.insert_comm _ cfg.A₁, set.insert_comm _ cfg.A₁, set.pair_comm],
+  exact set.subset_insert _ _
+end
+
+lemma A₂_ne_C : cfg.A₂ ≠ cfg.C :=
+begin
+  intro h,
+  have h₁ := cfg.sbtw_A_A₁_A₂,
+  rw h at h₁,
+  refine cfg.not_collinear_ABC _,
+  have hc : collinear ℝ ({cfg.A, cfg.B, cfg.C, cfg.A₁} : set Pt) :=
+    collinear_insert_insert_of_mem_affine_span_pair h₁.left_mem_affine_span
+      cfg.sbtw_B_A₁_C.left_mem_affine_span,
+  refine hc.subset (set.insert_subset_insert (set.insert_subset_insert _)),
+  rw set.singleton_subset_iff,
+  exact set.mem_insert _ _
+end
+
+lemma B₂_ne_B : cfg.B₂ ≠ cfg.B := by { rw ←symm_A₂, exact cfg.symm.A₂_ne_A }
+
+lemma B₂_ne_Q : cfg.B₂ ≠ cfg.Q := by { rw ←symm_A₂, exact cfg.symm.A₂_ne_P }
+
+lemma B₂_ne_A₂ : cfg.B₂ ≠ cfg.A₂ :=
+begin
+  intro h,
+  have hA : sbtw ℝ (cfg.triangle_ABC.points 1) cfg.A₁ (cfg.triangle_ABC.points 2) :=
+    cfg.sbtw_B_A₁_C,
+  have hB : sbtw ℝ (cfg.triangle_ABC.points 0) cfg.B₁ (cfg.triangle_ABC.points 2) :=
+    cfg.sbtw_A_B₁_C,
+  have hA' : cfg.A₂ ∈ line[ℝ, cfg.triangle_ABC.points 0, cfg.A₁] :=
+    sphere.second_inter_vsub_mem_affine_span _ _ _,
+  have hB' : cfg.A₂ ∈ line[ℝ, cfg.triangle_ABC.points 1, cfg.B₁],
+  { rw ←h, exact sphere.second_inter_vsub_mem_affine_span _ _ _ },
+  exact (sbtw_of_sbtw_of_sbtw_of_mem_affine_span_pair dec_trivial hA hB hA' hB').symm.not_rotate
+    cfg.sbtw_A_A₁_A₂.wbtw
+end
+
+lemma wbtw_B_Q_B₂ : wbtw ℝ cfg.B cfg.Q cfg.B₂ := cfg.sbtw_B_B₁_B₂.wbtw.trans_left cfg.wbtw_B_Q_B₁
+
+section oriented
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂ :
+  (2 : ℤ) • ∡ cfg.Q cfg.P cfg.A₂ = (2 : ℤ) • ∡ cfg.B cfg.A cfg.A₂ :=
```

Comment:
This `(2 : ℤ) • x = (2 : ℤ) • y` statement seems to come up all over the place, yet I never remember seeing it on paper.
Is it worth giving this notion its own name?

Reply:
See the above remark (in a conversation Kevin marked resolved and I've now unresolved) about the possibility of defining `real.angle.eq_mod_pi` so we could say `x ≡[π] y` or similar instead. I expect that would be straightforward to add, but also (as noted above) harder to rewrite with because special-purpose rewrite lemmas would be needed rather than just rewriting directly with an `eq_mod_pi` hypothesis. (Though I suppose other relations in mathlib such as `modeq` already have that "harder to rewrite with" property.)